### PR TITLE
Supporting index sort and length on Migration Engine

### PR DESCRIPTION
**Pull request by pimeys**
Closes: https://github.com/prisma/prisma/issues/9577

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -2617,12 +2618,9 @@ dependencies = [
 
 [[package]]
 name = "object"
-version = "0.26.2"
+version = "0.24.0"
```
Comment (tomhoule) [inline Cargo.lock:2621]: There's been a bunch of downgrades. Do you know why?

File: libs/sql-ddl/src/common.rs [hunk 2]
```diff
@@ -37,3 +40,46 @@ where
...
+
+#[derive(Debug, Default)]
+pub struct IndexColumn<'a> {
+    pub name: Cow<'a, str>,
+    pub length: Option<u32>,
+    pub sort_order: Option<SortOrder>,
+}
+
+impl<'a> IndexColumn<'a> {
+    pub fn new(name: &'a str) -> Self {
```
Comment (tomhoule) [inline libs/sql-ddl/src/common.rs:79]: The design of `sql-ddl` idea was "only plain structs, no methods". I'm not convinced we still want sql-ddl anymore, so let's not care for now.

File: libs/sql-ddl/src/mysql.rs [hunk 3]
```diff
@@ -193,7 +193,24 @@ impl Display for CreateIndex<'_> {
             table_name = self.on.0,
         )?;
 
-        self.on.1.iter().map(|s| Ident(s)).join(", ", f)?;
+        self.on
+            .1
+            .iter()
+            .map(|s| {
+                let mut rendered = format!("{}", Ident(&s.name));
```
Comment (tomhoule) [inline libs/sql-ddl/src/mysql.rs:200]: this can be `Ident(&s.name).to_string()`.

File: libs/sql-ddl/src/postgres.rs [hunk 4]
```diff
@@ -386,6 +400,34 @@ mod tests {
...
+            index_name: "meow_idx".into(),
+            table_reference: "Cat".into(),
+            columns,
+        };
+
+        assert_eq!(
+            create_index.to_string(),
+            "CREATE UNIQUE INDEX \"meow_idx\" ON \"Cat\"(\"name\" ASC, \"age\" DESC)"
+        )
+    }
```
Comment (tomhoule) [inline libs/sql-ddl/src/postgres.rs:429]: :+1:

File: migration-engine/connectors/mongodb-migration-connector/src/schema_calculator.rs [hunk 5]
```diff
@@ -1,9 +1,10 @@
 use crate::schema::MongoSchema;
-use datamodel::Datamodel;
+use datamodel::{common::preview_features::PreviewFeature, Datamodel};
+use enumflags2::BitFlags;
 use mongodb::bson::{Bson, Document};
 
 /// Datamodel -> MongoSchema
-pub(crate) fn calculate(datamodel: &Datamodel) -> MongoSchema {
+pub(crate) fn calculate(datamodel: &Datamodel, preview_features: BitFlags<PreviewFeature>) -> MongoSchema {
```
Comment (tomhoule) [inline migration-engine/connectors/mongodb-migration-connector/src/schema_calculator.rs:7]: In the calculator, the preview feature will already have been feature gated, so we will only ever get `DESC` in the datamodel if it's enabled. I think we don't need that new parameter.

File: migration-engine/connectors/mongodb-migration-connector/src/client_wrapper.rs [hunk 6]
```diff
@@ -67,8 +69,16 @@ impl Client {
...
-                let path = index.keys;
-                schema.push_index(collection_id, name, is_unique, path);
+                let path = if preview_features.contains(PreviewFeature::ExtendedIndexes) {
+                    index.keys
+                } else {
+                    index.keys.iter().fold(Document::new(), |mut acc, (k, _)| {
+                        acc.insert(k, Bson::Int32(1));
+                        acc
+                    })
+                };
```
Comment (tomhoule) [inline migration-engine/connectors/mongodb-migration-connector/src/client_wrapper.rs:79]: I think we're missing a test about changing the direction of an index (asc -> desc or the other way). I don't think we're diffing that in this PR.

File: migration-engine/connectors/mongodb-migration-connector/tests/migrations/scenarios/indexes_can_be_created_descending/schema.prisma [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+   @@index([firstName, lastName(sort: Desc), age])
+ }
+
+ model Post {
+   id Int @id
+   title String
+
+   @@index([title(sort: Desc)])
+ }
+
```
Comment (tomhoule) [inline migration-engine/connectors/mongodb-migration-connector/tests/migrations/scenarios/indexes_can_be_created_descending/schema.prisma:21]: :+1:

**Review by tomhoule — COMMENTED** [review]
First read: mongo migration connector. Going to read the SQL part now.

File: migration-engine/migration-engine-tests/tests/migrations/id.rs [hunk 8]
```diff
@@ -1,6 +1,203 @@
...
+          id Int @id
+        }}
+    "#, api.datasource_block()};
+
+    api.schema_push(&dm).force(true).send().assert_green();
+
+    api.assert_schema().assert_table("A", |table| {
+        table.assert_pk(|pk| pk.assert_column("id", |attr| attr.assert_sort_order(SQLSortOrder::Asc)))
+    });
+}
```
Comment (tomhoule) [inline migration-engine/migration-engine-tests/tests/migrations/id.rs:200]: :+1:

File: migration-engine/migration-engine-tests/src/assertions.rs [hunk 9]
```diff
@@ -710,4 +753,18 @@ impl<'a> IndexAssertion<'a> {
...
+    {
+        let col = self.0.columns.iter().find(|i| i.name == column_name).unwrap();
+
+        f(IndexColumnAssertion {
+            sort_order: col.sort_order,
+            length: col.length,
+        });
+
+        self
+    }
```
Comment (tomhoule) [inline migration-engine/migration-engine-tests/src/assertions.rs:769]: :100:

**Review by tomhoule — COMMENTED** [review]
The SQL part looks good, no nitpick :+1:

File: migration-engine/connectors/sql-migration-connector/src/sql_renderer/mssql_renderer.rs [hunk 10]
```diff
@@ -156,7 +166,24 @@ impl SqlRenderer for MssqlFlavour {
...
+                .columns
+                .iter()
+                .map(|col| {
+                    let mut rendered = format!("{}", self.quote(col.name()));
+
+                    if let Some(sort_order) = col.sort_order {
+                        rendered.push(' ');
+
+                        match sort_order {
+                            SQLSortOrder::Asc => rendered.push_str("ASC"),
```
Comment (do4gr) [inline migration-engine/connectors/sql-migration-connector/src/sql_renderer/mssql_renderer.rs:179]: can we unify the different Option<SortOrder> -> str  methods?

File: migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/table.rs [hunk 11]
```diff
@@ -157,10 +157,16 @@ impl<'schema, 'b> TableDiffer<'schema, 'b> {
 
 /// Compare two SQL indexes and return whether they only differ by name.
 fn indexes_match(first: &IndexWalker<'_>, second: &IndexWalker<'_>) -> bool {
-    let left_cols = first.column_names();
-    let right_cols = second.column_names();
+    let left_cols = first.columns();
+    let right_cols = second.columns();
 
     left_cols.len() == right_cols.len()
```
Comment (do4gr) [inline migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/table.rs:163]: This needs to start diffing on length and sort differences. I'm pretty sure those will only be populated when the preview flag is set, but we should test that.

**Review by do4gr — APPROVED** [review]

**Review by tomhoule — APPROVED** [review]

**Pull request merged**